File: child-app/app/src/main/java/com/familysafety/child/utils/EncryptionUtils.kt
package com.familysafety.child.utils

import android.util.Base64
import javax.crypto.Cipher
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec
import java.security.SecureRandom

object EncryptionUtils {
    private const val ALGORITHM = "AES/GCM/NoPadding"
    private const val TAG_LENGTH_BIT = 128
    private const val IV_LENGTH_BYTE = 12

    fun encrypt(data: String, secretKey: String): String {
        val cipher = Cipher.getInstance(ALGORITHM)
        val iv = ByteArray(IV_LENGTH_BYTE)
        SecureRandom().nextBytes(iv)
        
        val keySpec = SecretKeySpec(secretKey.toByteArray(), "AES")
        val gcmSpec = GCMParameterSpec(TAG_LENGTH_BIT, iv)
        
        cipher.init(Cipher.ENCRYPT_MODE, keySpec, gcmSpec)
        val cipherText = cipher.doFinal(data.toByteArray())
        
        val combined = iv + cipherText
        return Base64.encodeToString(combined, Base64.DEFAULT)
    }

    fun decrypt(encryptedData: String, secretKey: String): String {
        val combined = Base64.decode(encryptedData, Base64.DEFAULT)
        val iv = combined.sliceArray(0 until IV_LENGTH_BYTE)
        val cipherText = combined.sliceArray(IV_LENGTH_BYTE until combined.size)
        
        val cipher = Cipher.getInstance(ALGORITHM)
        val keySpec = SecretKeySpec(secretKey.toByteArray(), "AES")
        val gcmSpec = GCMParameterSpec(TAG_LENGTH_BIT, iv)
        
        cipher.init(Cipher.DECRYPT_MODE, keySpec, gcmSpec)
        return String(cipher.doFinal(cipherText))
    }
}

File: parent-app/src/utils/EncryptionUtils.ts
import { Buffer } from 'buffer';
import * as CryptoJS from 'crypto-js';

export const EncryptionUtils = {
  encrypt: (data: string, secretKey: string) => {
    const iv = CryptoJS.lib.WordArray.random(12);
    const encrypted = CryptoJS.AES.encrypt(data, CryptoJS.enc.Utf8.parse(secretKey), {
      iv: iv,
      mode: CryptoJS.mode.GCM,
      padding: CryptoJS.pad.NoPadding
    });
    
    return iv.toString(CryptoJS.enc.Base64) + encrypted.toString();
  },

  decrypt: (encryptedData: string, secretKey: string) => {
    const iv = CryptoJS.enc.Base64.parse(encryptedData.substring(0, 16)); // Base64 12 bytes
    const cipherText = encryptedData.substring(16);
    
    const decrypted = CryptoJS.AES.decrypt(cipherText, CryptoJS.enc.Utf8.parse(secretKey), {
      iv: iv,
      mode: CryptoJS.mode.GCM,
      padding: CryptoJS.pad.NoPadding
    });
    
    return decrypted.toString(CryptoJS.enc.Utf8);
  }
};
